### PR TITLE
feature: allow opam files in opam/

### PR DESCRIPTION
**Pull request by rgrinberg**
When we set (opam_file_location inside_opam_directory), we are allowed
to put all the opam file in the opam/ directory relative to the project.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Comment by Alizter** [comment]
Would we want this to be more configurable? There might be users that don't want an `opam/` directory but would like to give their own.

**Comment by rgrinberg** [comment]
> There might be users that don't want an opam/ directory but would like to give their own.

This is the only directory that opam understands for the purposes of pinning

File: src/dune_rules/dune_project.ml [hunk 1]
```diff
@@ -243,6 +245,10 @@ let to_dyn
     ; ("strict_package_deps", bool strict_package_deps)
     ; ("cram", bool cram)
     ; ("expand_aliases_in_sandbox", bool expand_aliases_in_sandbox)
+    ; ( "opam_file_lcoation"
```
Comment (emillon) [inline src/dune_rules/dune_project.ml:248]:
```suggestion
    ; ( "opam_file_location"
```

File: src/dune_rules/opam_create.ml [hunk 2]
```diff
@@ -281,7 +282,23 @@ let generate project pkg ~template =
     | None -> ""
     | Some (_, s) -> s)
 
-let add_rule sctx ~project ~pkg =
+let add_alias_rule sctx ~project ~pkg =
```
Comment (anmonteiro) [inline src/dune_rules/opam_create.ml:285]: If I'm reading this correctly, then we still emit the `.opam` files in the repository root inside the build folder?

Comment (anmonteiro) [reply src/dune_rules/opam_create.ml:285]: to answer my own question: we don't emit `*.opam` files in `opam/`, we just expect them to be there.

**Review by anmonteiro — APPROVED** [review]
LGTM. 

The only question I have, for which I'm unsure of the tradeoffs, is:

the difference between telling gen_rules about this `opam/` directory as an automatic_subdir vs making `source_tree` aware of it.

**Comment by rgrinberg** [comment]
> vs making source_tree aware of it.

I'm in the process of moving away as much logic as possible from Source_tree. The problem with all this partial evaluation of dune files is that makes it impossible to make anything generated by rules or interpreted differently based on the context or profile.

**Comment by Alizter** [comment]
@emillon This does not appear to be documented.

**Comment by emillon** [comment]
indeed. do you want to submit a PR? I think it's just a matter of adding the new field in `dune-files.rst`.

**Comment by ejgallego** [comment]
Also a CHANGES entry for this seems to be missing.

**Comment by ejgallego** [comment]
The lack of changes file means that we can't document to users what dune version they need in order to use this feature.

**Comment by Alizter** [comment]
If its any consolation, it was introduced in 3.8.

**Pull request merged**